# 예측모델  V.3-1 (Score 0.63636) 0.66028 .. 똑같은거 다시 제출했는데 다르게나오긴하네

기존에 주어진 컬럼들 중

- **`Pclass`** 와 **`Sex`** 만 이용하여 생존 여부를 예측

- **`Fare`**  를 `Categorize` 하여 추가하기 ( + Fare 의 Null Data 처리 -> Pclass 컬럼을 이용하여 예측 )

- `NaN` 값이 있는  **`Age`** 를 `Categorize` 하여 추가하기  (NaN 데이터는 제거)

# csv 파일 읽기

In [41]:
import pandas as pd

df_train = pd.read_csv('./input/train.csv')
df_test = pd.read_csv('./input/test.csv')


In [2]:
# df_train 데이터 엿보기 (default : 5)
df_train.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S

In [3]:
df_test.head()

PassengerId  Pclass                                          Name     Sex  \
0          892       3                              Kelly, Mr. James    male   
1          893       3              Wilkes, Mrs. James (Ellen Needs)  female   
2          894       2                     Myles, Mr. Thomas Francis    male   
3          895       3                              Wirz, Mr. Albert    male   
4          896       3  Hirvonen, Mrs. Alexander (Helga E Lindqvist)  female   

    Age  SibSp  Parch   Ticket     Fare Cabin Embarked  
0  34.5      0      0   330911   7.8292   NaN        Q  
1  47.0      1      0   363272   7.0000   NaN        S  
2  62.0      0      0   240276   9.6875   NaN        Q  
3  27.0      0      0   315154   8.6625   NaN        S  
4  22.0      1      1  3101298  12.2875   NaN        S

In [4]:
df_train.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 12 columns):
PassengerId    891 non-null int64
Survived       891 non-null int64
Pclass         891 non-null int64
Name           891 non-null object
Sex            891 non-null object
Age            714 non-null float64
SibSp          891 non-null int64
Parch          891 non-null int64
Ticket         891 non-null object
Fare           891 non-null float64
Cabin          204 non-null object
Embarked       889 non-null object
dtypes: float64(2), int64(5), object(5)
memory usage: 83.6+ KB


In [5]:
df_test.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 418 entries, 0 to 417
Data columns (total 11 columns):
PassengerId    418 non-null int64
Pclass         418 non-null int64
Name           418 non-null object
Sex            418 non-null object
Age            332 non-null float64
SibSp          418 non-null int64
Parch          418 non-null int64
Ticket         418 non-null object
Fare           417 non-null float64
Cabin          91 non-null object
Embarked       418 non-null object
dtypes: float64(2), int64(4), object(5)
memory usage: 36.0+ KB


In [6]:
df_train.describe()

PassengerId    Survived      Pclass         Age       SibSp  \
count   891.000000  891.000000  891.000000  714.000000  891.000000   
mean    446.000000    0.383838    2.308642   29.699118    0.523008   
std     257.353842    0.486592    0.836071   14.526497    1.102743   
min       1.000000    0.000000    1.000000    0.420000    0.000000   
25%     223.500000    0.000000    2.000000   20.125000    0.000000   
50%     446.000000    0.000000    3.000000   28.000000    0.000000   
75%     668.500000    1.000000    3.000000   38.000000    1.000000   
max     891.000000    1.000000    3.000000   80.000000    8.000000   

            Parch        Fare  
count  891.000000  891.000000  
mean     0.381594   32.204208  
std      0.806057   49.693429  
min      0.000000    0.000000  
25%      0.000000    7.910400  
50%      0.000000   14.454200  
75%      0.000000   31.000000  
max      6.000000  512.329200

# 유의미해 ?  ? ? ?

In [7]:
df_train['CategoricalFare'] = pd.qcut(df_train['Fare'], 4)
df_test['CategoricalFare'] = pd.qcut(df_train['Fare'], 4)

print( df_train[['Fare','CategoricalFare']].head())

      Fare  CategoricalFare
0   7.2500   (-0.001, 7.91]
1  71.2833  (31.0, 512.329]
2   7.9250   (7.91, 14.454]
3  53.1000  (31.0, 512.329]
4   8.0500   (7.91, 14.454]


In [8]:
print( df_train[['CategoricalFare','Survived']].groupby('CategoricalFare').mean() )

                 Survived
CategoricalFare          
(-0.001, 7.91]   0.197309
(7.91, 14.454]   0.303571
(14.454, 31.0]   0.454955
(31.0, 512.329]  0.581081


In [9]:
df_train['CategoricalFare'] = pd.qcut(df_train['Fare'], 4)

In [10]:
# df_train['CategoricalAge'] = pd.qcut(df_train['Age'], 15)

# print( df_train[['CategoricalAge','Survived']].groupby('CategoricalAge').mean() )

In [11]:
#Pclass, Sex, Fare, Age 컬럼만 가지고 감

df_train = df_train[['Survived','Pclass','Sex','Fare', 'Age']]
df_test = df_test[['Pclass','Sex','Fare', 'Age']]

df_train.head()

Survived  Pclass     Sex     Fare   Age
0         0       3    male   7.2500  22.0
1         1       1  female  71.2833  38.0
2         1       3  female   7.9250  26.0
3         1       1  female  53.1000  35.0
4         0       3    male   8.0500  35.0

# Null Check

In [12]:
for col in df_train.columns:
    msg = 'column: {:>10}\t Percent of NaN value: {:.2f}%'.format(col, 100 * (df_train[col].isnull().sum() / df_train[col].shape[0]))
    print(msg)

column:   Survived	 Percent of NaN value: 0.00%
column:     Pclass	 Percent of NaN value: 0.00%
column:        Sex	 Percent of NaN value: 0.00%
column:       Fare	 Percent of NaN value: 0.00%
column:        Age	 Percent of NaN value: 19.87%


# Null 데이터 처리

In [13]:
df_train = df_train.dropna()
df_test = df_test.fillna(0)

In [14]:
df_train.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 714 entries, 0 to 890
Data columns (total 5 columns):
Survived    714 non-null int64
Pclass      714 non-null int64
Sex         714 non-null object
Fare        714 non-null float64
Age         714 non-null float64
dtypes: float64(2), int64(2), object(1)
memory usage: 33.5+ KB


In [15]:
df_test.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 418 entries, 0 to 417
Data columns (total 4 columns):
Pclass    418 non-null int64
Sex       418 non-null object
Fare      418 non-null float64
Age       418 non-null float64
dtypes: float64(2), int64(1), object(1)
memory usage: 13.1+ KB


`dropna()` 를 이용하여 

**`Age`** 가 `NaN` 인 행을 삭제하였더니

`df_train` 은 714행을 가지는 df가 되었다.

# Data Cleaning

## 1) Sex

In [16]:
df_train['Sex'] = df_train['Sex'].map({'female': 0, 'male': 1})
df_test['Sex'] = df_test['Sex'].map({'female': 0, 'male': 1})





## 2) Fare

In [17]:
def category_fare(x):
    if x <= 7.91:
        return 0
    elif x <= 14.454:
        return 1
    elif x <= 31.0:
        return 2
    else:
        return 3

df_train['Fare'] = df_train['Fare'].apply(category_fare)    
df_test['Fare'] = df_test['Fare'].apply(category_fare)

## 3) Age

In [25]:
df_train['CategoricalAge'] = pd.cut(df_train['Age'], 5)
print (df_train[['CategoricalAge', 'Survived']].groupby(['CategoricalAge'], as_index=False).mean())

     CategoricalAge  Survived
0    (0.34, 16.336]  0.550000
1  (16.336, 32.252]  0.369942
2  (32.252, 48.168]  0.404255
3  (48.168, 64.084]  0.434783
4    (64.084, 80.0]  0.090909


In [22]:
def category_age(x):
    if x <= 10:
        return 0
    elif x <= 20:
        return 1
    elif x <= 30:
        return 2
    elif x <= 40:
        return 3
    elif x <= 50:
        return 4
    elif x<= 60:
        return 5
    elif x <= 70:
        return 6
    else:
        return 7
        

     CategoricalAge  Survived
0    (0.34, 16.336]  0.550000
1  (16.336, 32.252]  0.369942
2  (32.252, 48.168]  0.404255
3  (48.168, 64.084]  0.434783
4    (64.084, 80.0]  0.090909


In [ ]:
df_train['Age'] = df_train['Age'].apply(category_age)
df_test['Age'] = df_test['Age'].apply(category_age)

# 원래 float 형이었는데, int 형으로 자동으로 바껴지네..

In [ ]:
df_train.head()

# 모델 만들기

In [ ]:
from sklearn.ensemble import RandomForestClassifier 
from sklearn import metrics 
from sklearn.model_selection import train_test_split


In [ ]:
#학습시킬 생존 여부 값
target_label = df_train['Survived'].values

# 학습시킬 환경 값
X_train = df_train.drop('Survived', axis=1).values

# 테스트 할 환경 값
X_test = df_test.values

In [ ]:
print( X_train )

In [ ]:
X_tr, X_vld, y_tr, y_vld = train_test_split(X_train, target_label, test_size=0.3, random_state=2018) # 입력값, 결과값, test_size, random_state ??

In [ ]:
# 모의고사

model = RandomForestClassifier()
model.fit(X_tr, y_tr)
prediction = model.predict(X_vld)

print('총 {}명 중 {:.2f}% 정확도로 생존을 맞춤'.format(y_vld.shape[0], 100 * metrics.accuracy_score(prediction, y_vld)))

In [ ]:
from pandas import Series
import matplotlib.pyplot as plt

plt.style.use('seaborn')

plt.figure(figsize=(8, 8))

Series(model.feature_importances_, index=df_test.columns).sort_values(ascending=True).plot.barh()

plt.xlabel('Feature importance')
plt.ylabel('Feature')

plt.show()

In [ ]:
submission = pd.read_csv('./input/gender_submission.csv')

submission.head()

In [ ]:
X_test

In [ ]:
submission['Survived'] = model.predict(X_test)

submission.head(10)

In [ ]:
model.predict(X_test)

In [ ]:
for col in df_train.columns:
    msg = 'column: {:>10}\t Percent of NaN value: {:.2f}%'.format(col, 100 * (df_train[col].isnull().sum() / df_train[col].shape[0]))
    print(msg)

In [ ]:
df_test.info()

In [ ]:
submission['Survived'].mean()

# 제출용 csv 파일 만들기

In [ ]:
submission.to_csv('./output/v3-1.csv', index=False)